src/libpmemobj/cuckoo.c

```diff
@@ -160,47 +228,88 @@ cuckoo_insert_try(struct cuckoo *c, struct cuckoo_slot *src)
 }
 
 /*
+ * cuckoo_update_on_grow -- (internal) updates the observable state of the table
+ * to point to the newly rehash internal tab
+ *
+ * The only two fields which matter in this regard, are `tab` and `size`,
+ * as cuckoo_get does only access these.
+ *
+ * XXX clean up atomic operations!
+ */
+static void
+cuckoo_update_on_grow(struct cuckoo *observable_state, struct cuckoo *new_state)
+{
+#ifdef _MSC_VER
+	observable_state->tab = new_state->tab;
+#else
+	__atomic_store_n(&observable_state->tab, new_state->tab,
+	    __ATOMIC_RELEASE);
+#endif
+	/*
+	 * At this point, other threads might be accessing the
+	 * newly built table, using the old size. Since the size
+	 * of this table never decreases, such accesses always
+	 * refer to accessible memory. The values read from
+	 * that memory are unspecified, but this is indicated
+	 * using the status field of the cuckoo struct.
+	 */
+#ifdef _MSC_VER
+	observable_state->size = new_state->size;
+#else
+	__atomic_store_n(&observable_state->size, new_state->size,
+	    __ATOMIC_RELEASE);
+#endif
+}
+
+/*
  * cuckoo_grow -- (internal) rehashes the table with twice the size
+ *
+ * The rehashed table is not visible to other threads while it is under
+ * construction. This routine only modifies anything observable to other threads
+ * once it reaches a point where no more failures are possible. The
+ * modifications are applied to the actual visible state of the cuckoo in a
+ * specific order, to make sure other threads accessing it do not dereference
+ * invalid pointers.
  */
 static int
-cuckoo_grow(struct cuckoo *c)
+cuckoo_grow(struct cuckoo *observable)
 {
-	size_t oldsize = c->size;
-	struct cuckoo_slot *oldtab = c->tab;
+	struct cuckoo c = *observable;
+	if (c.deferred_deallocations) {
+		/* Make sure there is enough space to hold pointers for later */
+		size_t old_tabs_nsize = sizeof(*c.tab) * (c.old_tab_count + 1u);
+		c.old_tabs = Realloc(c.old_tabs, old_tabs_nsize);
```

Comment:
leak

Reply:
If the reallocation was not successfull, the observable state did not change (the original pointer is still there).
If the reallocation was successfull, the pointer will be updated in the observable state (and should access old_tabs meanwhile, as inserts aren't lock-free).

Reply:
Fine, but this is really not apparent. Add a comment please.